Design_Doc_Examples/RAG_Q&A_for collaborative_work_platform.md

```diff
@@ -136,16 +136,110 @@ No ideas
-    1. Establishing a baseline is a critical first step in machine learning system design, serving as a simple, operational starting point for iterative improvement.
-    2. The choice of baseline should be guided by a trade-off between desired accuracy and the effort required for development, with simplicity often providing significant advantages in terms of robustness, scalability, and interpretability.
-    3. In deep learning applications, leveraging pretrained models or training simple models from scratch can provide effective baselines, with the choice influenced by the specific requirements and constraints of the project.
-    4. Continuous evaluation and comparison against the baseline are essential for guiding the development process, ensuring that complexity is added only when it yields proportional benefits in performance.
+#### Document Extraction Process
+
+Considering the minimal variability among the documents and the sufficient coverage of most cases within the current dataset, it is recommended to implement an in-house document extraction pipeline. This pipeline should consist of:
+
+1. File Type Handler: Differentiate and handle file types accordingly, as PDFs and images may require additional processing steps.
+2. Text Extraction: Deploy a customized OCR solution designed to handle non-text elements.
+3. Text Preprocessing: Remove unwanted characters, whitespace, or any artifacts.
+4. Markdown Formatting: Ensure that the extracted content is formatted correctly according to markdown standards.
+5. Error Management & Spell Checking: Integrate an error handler and a spell checker to maintain data quality.
+
+#### Retrieval-Augmented Generation Framework
+
+The Retrieval-Augmented Generation (RAG) framework can be broken down into two main components:
+
+- Retrieval
+- Augmented Generation
+
+Augmented Generation is a recent advancement, while the concept of document retrieval is something that has been with us since the emergine of web search. While there is little to no sense in building the second part using solutions other than LLMs, it might make sense to implement a simple baseline for the retrieval.
+
+#### Retrieval: Sparse Encoded Retrieval Baseline
+
+Objectives:
+- Create a robust baseline with minimal effort.
+- Validate the hypothesis that an enhanced search capability is beneficial.
+- Gather a dataset based on retrieval, incorporating both implicit and explicit feedback for future refinement.
+
+Applicability:
+This covers use case `1a`. The solution is not applicable to the use cases `1na` and `2na`, thus also addressed.
+
+The system enables content search within documents using the BM25 algorithm.
+
+Components:
+1. Preprocessing Layer
+    - Tokenizes input data
+    - Filters out irrelevant content
+    - Applies stemming / lemmatization
+2. Indexing Layer
+    - Maintains a DB-represented corpus
+    - Creates indexes for Term Frequency (TF) and Inverse Document Frequency (IDF)
+3. Inference Layer
```

Comment:
@tyomj  will encoding/retrieval be happening on the Token level or else?

Reply:
Yes, essentially, the preprocessing layer (which operates on the token level where the token is a stemmed char sequence) is utilized during inference time for the query as well. I'll make this explicit to avoid any confusion.